pandas/util/_decorators.py

```diff
@@ -247,6 +247,36 @@ def wrapper(*args, **kwargs) -> Callable[..., Any]:
     return decorate
 
 
+def doc(*args: Union[str, Callable], **kwargs: str) -> Callable:
+    """
+    A decorator take docstring templates, concatenate them and perform string
+    substitution on it.
+
+    This decorator should be robust even if func.__doc__ is None.
+    """
+
+    def decorator(func: Callable) -> Callable:
+        @wraps(func)
+        def wrapper(*args, **kwargs) -> Callable:
+            return func(*args, **kwargs)
+
+        templates = [func.__doc__ if func.__doc__ else ""]
+        for arg in args:
+            if isinstance(arg, str):
+                templates.append(arg)
+            elif hasattr(arg, "_docstr_template"):
+                templates.append(arg._docstr_template)  # type: ignore
+            elif arg.__doc__:
+                templates.append(arg.__doc__)
+
+        wrapper._docstr_template = "".join(dedent(t) for t in templates)  # type: ignore
+        wrapper.__doc__ = wrapper._docstr_template.format(**kwargs)  # type: ignore
```

Comment:
Yes, for current approach, we have to use `{{` and `}}` for `{` and `}`. I have a slight preference to keep it that way because I believe it can help keep behavior consistent and make it easier to use.

There are two kinds of docstring, in my opinion. We might name one "doc decorator friendly" and the other "doc decorator indeterminate":
1. The "doc decorator friendly" means the docstring has been modified to use `format`, such as using `{{` and `}}` for `{` and `}`.  We can safely put the function into a doc decorator without checking formatting details. 
2. For "doc decorator indeterminate", it may have `{` or `}` there. If we want to use it in doc decorator, we might need to convert it to "doc decorator friendly" first.

For the current approach, all template was guaranteed to be formatted before putting it in `__doc__`. Therefore, if there's a doc decorator, we can easily tell the function is "doc decorator friendly". In addition, the user won't have to worry about using `{{` and `}}` or `{` and `}` in the docstring, because of the consistent behavior.